packages/admin-panel-server/src/viz-builder/extractors/DraftDashboardItemExtractor.ts

```diff
@@ -0,0 +1,16 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { DashboardVisualisationExtractor } from './DashboardVisualisationExtractor';
+
+export class DraftDashboardItemExtractor extends DashboardVisualisationExtractor {
```

Comment:
Initially, I was thinking of just having 1 extractor `DraftVisualisationExtractor` class that will be responsible for extracting both report/dashboard_item from the visualisation object. However, there is 1 case that we only need to extract the report part. So I split it up into 2 extractor classes, 1 for report and 1 for dashboard_item. However, I do have a slight feeling that I may have over engineered it...

Reply:
Yep I think I agree with that. I still think we need inheritance here just because we will need different validation between  saving a draft version and saving a published version (draft version can still miss some fields, whereas published version we will have to validate everything). So I'd imagine the inhritance model would look like below:

```
                       AbstractDashboardVisualisation
                       //                         \\
                    //                               \\
                 //                                     \\
               V                                           V
DraftDashboardVisualisation                 PublishedDashboardVisualisation
```

What do you reckon?

Reply:
I think that should work 👍 Sometimes I get a bit worried about employing inheritance since it can often lead to additional complexity, but what you describe does seem pretty clean. Basically the only difference would be the validation logic right?

Reply:
Definitely understand your concern about inheritance adding additional complexity 👍 . Yep I think validation is the main difference.